### PR TITLE
INT-B-20522 SIT Validation

### PR DESCRIPTION
**Pull request by loganwc**
## Summary
Prime cannot set Payment Start date earlier than SIT entry date.

## Testing Instructions

1. Create an HHG and move to prime queue
2. *Prime* Update shipment and add shipment weight
3. *Prime* Create SIT Service Item
4. *TOO* Approve SIT item
5. *Prime* Create payment request -> enter an earlier date for the SIT payment request than the SIT entry date
6. Get the following error: "is in a conflicting state cannot have payment date earlier than SIT Entry date"

Swagger instructions for testing first AC: (Use your own move and service item IDs...)

in DBeaver go to the `mto_service_items` table and use the following filter: `re_service_id  = '05eb6ff1-5cf6-4918-b887-8260dda6b9fe'::uuid`

Use a service item that *is* approved

post: /prime/v1/payment-requests
payload:
```
{
    "isFinal": false,
    "moveTaskOrderID": "d29240f5-30fc-4f3d-8b87-1ab7ee2929eb",
    "serviceItems": [
         {
                "id": "d15b2c8d-e89b-4c23-9d84-7adb7b5a1054",
                "params": [
                 {
                      "key": "WeightBilled",
                     "value": "1000"
                 },
                 {
                    "key": "SITPaymentRequestStart",
                    "value": "2029-12-31"
                 },
                 {
                    "key": "SITPaymentRequestEnd",
                    "value": "2029-12-31"
                 }
             ]
        }
     ],
    "pointOfContact": "myself"
}
```

### CODE REVIEW

File: pkg/services/payment_request/payment_request_creator.go [hunk 1]
```diff
@@ -142,6 +142,25 @@ func (p *paymentRequestCreator) CreatePaymentRequest(appCtx appcontext.AppContex
...
+				// check if service item param is for SIT Payment start/end
+				if param.IncomingKey == "SITPaymentRequestStart" || param.IncomingKey == "SITPaymentRequestEnd" {
+					paymentDate, err := time.Parse("2006-01-02", param.Value)
+					if err != nil {
+						errStr := "failed to parse " + param.IncomingKey
+						return apperror.NewBadDataError(errStr)
+					}
+
+					if paymentServiceItem.MTOServiceItem.SITEntryDate != nil && paymentServiceItem.MTOServiceItem.SITEntryDate.After(paymentDate) {
+						return apperror.NewConflictError(paymentRequestArg.ID, "cannot have payment date earlier than SIT Entry date")
```
Comment (pambecker) [inline pkg/services/payment_request/payment_request_creator.go:154]: Can you update this message to 'say earlier than or equal to SIT Entry date' please.

File: pkg/services/payment_request/payment_request_creator.go [hunk 2]
```diff
@@ -142,6 +142,25 @@ func (p *paymentRequestCreator) CreatePaymentRequest(appCtx appcontext.AppContex
...
+
+					}
+
+					// Check if payment date is later than SIT dates
```
Comment (pambecker) [inline pkg/services/payment_request/payment_request_creator.go:158]: Minor detail, but did you mean this comment to be above both if statements?  doesn't seem to make sense here after you already checked one of the sit dates.

File: pkg/services/payment_request/payment_request_creator.go [hunk 3]
```diff
@@ -142,6 +142,25 @@ func (p *paymentRequestCreator) CreatePaymentRequest(appCtx appcontext.AppContex
...
+					if paymentServiceItem.MTOServiceItem.SITDepartureDate != nil && paymentServiceItem.MTOServiceItem.SITDepartureDate.After(paymentDate) {
```
Comment (pambecker) [inline pkg/services/payment_request/payment_request_creator.go:159]: Is this a necessary check?  I don't see anything about the departure date in the backlog.  If so, we should update the backlog.

File: pkg/handlers/primeapi/payment_request.go [hunk 4]
```diff
@@ -127,6 +127,14 @@ func (h CreatePaymentRequestHandler) Handle(params paymentrequestop.CreatePaymen
 						zap.Any("payload", payload))
 					return paymentrequestop.NewCreatePaymentRequestBadRequest().WithPayload(payload), err
 				default:
+					if e.Error() == "cannot have a payment date earlier than SIT Entry date" {
```
Comment (pambecker) [inline pkg/handlers/primeapi/payment_request.go:130]: would the code ever get to this point?  seems like it hit the ConflictError above on line 103

File: pkg/services/payment_request/payment_request_creator.go (hunk 3 above)
Comment (loganwc) [reply pkg/services/payment_request/payment_request_creator.go:159]: Well when I first picked up this item you could set the departure date earlier than the entry date for some stupid reason so I checked both

File: pkg/services/payment_request/payment_request_creator.go (hunk 2 above)
Comment (loganwc) [reply pkg/services/payment_request/payment_request_creator.go:158]: See comment below

File: pkg/handlers/primeapi/payment_request.go (hunk 4 above)
Comment (loganwc) [reply pkg/handlers/primeapi/payment_request.go:130]: It did regularly in my testing

Comment (pambecker) [reply pkg/handlers/primeapi/payment_request.go:130]: Hmmm...mine is hitting the conflict case above.  maybe someone else can try?  I'll go back and check also

**Review by paulstonebraker — APPROVED** [review]

**Comment by paulstonebraker** [comment]
Testing instructions were stellar, easy approve

**Review by pambecker — APPROVED** [review]

**Pull request merged**